lib/pbench/server/filetree.py

```diff
@@ -0,0 +1,778 @@
+        (controller_dir / "TODO").mkdir(exist_ok=True)
+        return Controller(controller_dir, options.INCOMING, options.RESULTS, logger)
+
+    def create_tarball(self, dataset: Path) -> Tarball:
+        """
+        Create a new dataset under the controller, link it to the controller,
+        and return the new Tarball object.
+
+        Args:
+            dataset: Path to source tarball file
+
+        Returns:
+            Tarball object
+        """
+        tarball = Tarball.create(dataset, self)
+        self.tarballs[tarball.name] = tarball
+        return tarball
+
+    def link(self, dataset: Path, state: str):
+        """
+        Create a state link within the controller sub-tree.
+
+        Args:
+            dataset: Tarball path
+            state: State directory name (e.g., "TODO")
+        """
+        (self.state_dirs[state] / dataset.name).symlink_to(dataset)
+
+    def unpack(self, dataset: str):
+        """
+        Unpack a tarball into the INCOMING tree. Create the INCOMING controller
+        directory if necessary, along with the RESULTS tree link.
+
+        NOTE: This does not look at tarball metadata.log, and therefore cannot
+        consider the `prefix` and `user` metadata which, in 0.69, affect
+        (respectively) the directory path of the RESULTS tree link, and whether
+        an entry is added under a different USERS tree. It's unclear we want
+        to preserve either behavior for 0.72 (especially USERS, which we've
+        replaced with true users and ownership).
+
+        Args:
+            dataset: Name of the dataset to unpack
+        """
+        tarball = self.tarballs[dataset]
+        self.incoming.mkdir(exist_ok=True)
+        self.results.mkdir(exist_ok=True)
+        tarball.unpack(self.incoming, self.results)
+
+    def uncache(self, dataset: str):
+        """
+        The reverse of `unpack`, removing the RESULTS tree link and the
+        unpacked tarball contents from INCOMING.
+
+        Args:
+            dataset: Name of dataset to remove
+        """
+        tarball = self.tarballs[dataset]
+        results_link = self.results / tarball.name
+        if results_link.exists():
+            results_link.unlink()
+        if self.results.exists() and not list(self.results.iterdir()):
+            self.results.rmdir()
+        incoming_dir = self.incoming / tarball.name
+        if incoming_dir.is_dir():
+            shutil.rmtree(incoming_dir, ignore_errors=True)
+        if self.incoming.exists() and not list(self.incoming.iterdir()):
+            self.incoming.rmdir()
+
+    def delete(self, dataset: str):
+        """
+        Delete a dataset and remove it from the controller. This will also
+        remove any links to the dataset tarball from the controller's state
+        directories.
+
+        NOTE: This relies on FileTree.delete() having already called the
+        controller uncache; we can't completely clean up within the controller
+        scope.
```

Comment:
I suppose I could check whether the tarball name exists in the INCOMING or RESULTS trees, sure; it seemed more reasonable at the time to just make a note rather than wasting the time checking.